catalogd/internal/storage/localdir.go

```diff
@@ -20,95 +26,353 @@ import (
 	if err != nil {
-		return false
+		httpError(w, err)
+		return
 	}
-	if !file.Mode().IsRegular() {
-		// path is not valid content
-		return false
+	serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
 	}
-	return true
+	defer catalogFile.Close()
+
+	w.Header().Set("Last-Modified", catalogStat.ModTime().UTC().Format(TimeFormat))
+	switch checkIfModifiedSince(r, w, catalogStat.ModTime()) {
+	case condNone:
+		return
+	case condFalse:
+		w.WriteHeader(http.StatusNotModified)
+		return
+	case condTrue:
+	}
+
+	schema := r.URL.Query().Get("schema")
+	pkg := r.URL.Query().Get("package")
+	name := r.URL.Query().Get("name")
+
+	if schema == "" && pkg == "" && name == "" {
+		// If no parameters are provided, return the entire catalog (this is the same as /api/v1/all)
+		serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+		return
+	}
+	idx, err := s.getIndex(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	indexReader, ok := idx.Get(catalogFile, schema, pkg, name)
+	if !ok {
+		httpError(w, fs.ErrNotExist)
+		return
+	}
+	serveJsonLinesQuery(w, indexReader)
+}
+
+func (s *LocalDirV1) catalogData(catalog string) (*os.File, os.FileInfo, error) {
+	catalogFile, err := os.Open(catalogFilePath(s.catalogDir(catalog)))
+	if err != nil {
+		return nil, nil, err
+	}
+	catalogFileStat, err := catalogFile.Stat()
+	if err != nil {
+		return nil, nil, err
+	}
+	return catalogFile, catalogFileStat, nil
+}
+
+func httpError(w http.ResponseWriter, err error) {
+	var code int
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		code = http.StatusNotFound
+	case errors.Is(err, fs.ErrPermission):
+		code = http.StatusForbidden
+	case errors.Is(err, ErrInvalidParams):
+		code = http.StatusBadRequest
+	default:
+		code = http.StatusInternalServerError
+	}
+	http.Error(w, fmt.Sprintf("%d %s", code, http.StatusText(code)), code)
 }
 
-// filesOnlyFilesystem is a file system that can open only regular
-// files from the underlying filesystem. All other file types result
-// in os.ErrNotExists
-type filesOnlyFilesystem struct {
-	FS fs.FS
+func serveJsonLines(w http.ResponseWriter, r *http.Request, modTime time.Time, rs io.ReadSeeker) {
+	w.Header().Add("Content-Type", "application/jsonl")
+	http.ServeContent(w, r, "", modTime, rs)
 }
 
-// Open opens a named file from the underlying filesystem. If the file
-// is not a regular file, it return os.ErrNotExists. Callers are resposible
-// for closing the file returned.
-func (f *filesOnlyFilesystem) Open(name string) (fs.File, error) {
-	file, err := f.FS.Open(name)
+func serveJsonLinesQuery(w http.ResponseWriter, rs io.Reader) {
+	w.Header().Add("Content-Type", "application/jsonl")
+	_, err := io.Copy(w, rs)
```

Comment:
We should only copy data if it is a GET request. a HEAD request should be identical to GET except without the response body.